internal/proxy/proxy_test.go

```diff
@@ -0,0 +1,170 @@
+	}
+	for name, tc := range tests {
+		tc := tc
+		name := name
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			_, err := proxy.New(context.Background(), tc.http, tc.https, tc.ftp, tc.socks, tc.noProxy, tc.mode)
+			if tc.wantErr {
+				require.Error(t, err, "New should have failed but didn't")
+				return
+			}
+			require.NoError(t, err, "New failed but shouldn't have")
+		})
+	}
+}
+
+func TestApply(t *testing.T) {
+	t.Parallel()
+
+	envConfigPath := filepath.Join("etc", "environment.d", "99ubuntu-proxy.conf")
+
+	initialTime := time.Date(2001, 1, 1, 1, 1, 1, 1, time.UTC)
+
+	tests := map[string]struct {
+		http    string
+		https   string
+		ftp     string
+		socks   string
+		noProxy string
+		mode    string
+
+		existingDirs       []string
+		existingPerms      map[string]os.FileMode
+		prevContents       map[string]string
+		wantUnchangedFiles []string
+		dryRun             bool
+
+		wantErr bool
+	}{
+		"No options set":   {},
+		"HTTP option set":  {http: "http://example.com:8080"},
+		"Some options set": {http: "http://example.com:8080", https: "https://example.com:8080"},
+		"Some options set, environment parent directories are created": {
+			http: "http://example.com:8080", https: "https://example.com:8080", existingDirs: []string{},
+		},
+		"Some options set, environment file is unchanged": {
+			wantUnchangedFiles: []string{envConfigPath},
+			http:               "http://example.com:8080", https: "https://example.com:8080",
+			prevContents: map[string]string{envConfigPath: fmt.Sprintf(`%s
+HTTP_PROXY=http://example.com:8080
+http_proxy=http://example.com:8080
+HTTPS_PROXY=https://example.com:8080
+https_proxy=https://example.com:8080
+`, proxy.ConfHeader)},
+		},
+		"All options set":                              {http: "http://example.com:8080", https: "https://example.com:8080", ftp: "ftp://example.com:8080", socks: "socks://example.com:8080", noProxy: "localhost,127.0.0.1"},
+		"Password is escaped":                          {http: "http://username:p@$$:w0rd@example.com:8080"},
+		"Escaped password is not escaped again":        {http: "http://username:p%40$$%3Aw0rd@example.com:8080"},
+		"Domain username is escaped":                   {http: `http://EXAMPLE\bobsmith:p@$$:w0rd@example.com:8080`},
+		"Domain username without password is escaped":  {http: `http://EXAMPLE\bobsmith@example.com:8080`},
+		"Escaped domain username is not escaped again": {http: `http://EXAMPLE%5Cbobsmith@example.com:8080`},
+
+		"Proxy files not created when ran with dry-run": {dryRun: true},
+
+		// Error cases
+		"Error when environment config file is read-only":    {existingPerms: map[string]os.FileMode{envConfigPath: 0444}, prevContents: map[string]string{envConfigPath: "something"}, wantErr: true},
+		"Error when environment directory is a file":         {existingDirs: []string{"etc/"}, prevContents: map[string]string{filepath.Dir(envConfigPath): "something"}, wantErr: true},
```

Comment:
What you want to trigger here is can’t write to environment directory, so I think the test description should reflect this.